src/core/pointcloud/qgspointcloudlayerrenderer.cpp

```diff
@@ -172,13 +174,127 @@ bool QgsPointCloudLayerRenderer::forceRasterRender() const
   return true;
 }
 
-QList<IndexedPointCloudNode> QgsPointCloudLayerRenderer::traverseTree( const QgsPointCloudIndex *pc,
+QVector<QMap<QString, QVariant>> QgsPointCloudLayerRenderer::identify( const QgsGeometry &geometry, const QgsIdentifyContext &identifyContext )
```

Comment:
I would suggest splitting this function up into two parts:

1. A method in QgsPointCloudDataProvider which accepts a QgsBox3d extent (which must be in the layer's crs), a maximum screen error, and a maximum number of point to identify (or -1 for "every point"). This method would then contain all the index specific logic, and NOT handle any renderer related concepts.
2. A renderedPointsInBounds() method in QgsPointCloudRenderer (not QgsPointCloudLayerRenderer), which calls the method from (1) to get ALL the points from the data provider, and then does the point size and shape handling you've got here. This method would need arguments for the QgsBox3d, a render context, and the identify context.

This would be cleaner, and the main benefit is that we would then have a nice method for plugins to use to also perform a "give me the points at this position" search on a point cloud layer.

Reply:
My reason behind implementing this in the layer renderer is because with have access to rendering informations like the screen errors and the classification stuff.
How can we deal with the level of detail and classification if we move it to the data provider ?